### DISCUSSION TITLE
Understanding QR Code Format of LEDVANCE products (for Zigbee 3.0 QR-Code Commissioning with Install Codes)

### DISCUSSION THREAD

**Discussion by clemwo** [Koenkk/zigbee2mqtt#8498] [Devices]
Hi, this question is not directly related to the current version of Zigbee2MQTT but IMO this could be relevant in the future (when/if Zigbee 3.0 QR-Code commissiong gets added) so I think its fine to post it here for further opinions.   
Here's the deal: 
I have a LEDVANCE Smart+ Lighting Classic E27 Tunable White I added to a Bosch Smart Home Controller for testing purposes in light of my bachelor thesis which has the topic of the security of Zigbee implementations.  
The Bosch Smart Home Controller uses a so called "Installation Code" for better security of the commissioning process. The install code is encoded in a QR-Code thats printed on the side of the product. If you havent heard about it, here is a great resource on it: 
https://www.silabs.com/documents/public/application-notes/an1089-using-installation-codes-with-zigbee-devices.pdf
In order to add a device to the Bosch Smart Home system you just have to scan the QR-Code and the commissioning takes place which works great and I love the fact that the network key doesnt get encrypted with a well-known key (which happens during the "classic" commissioning process that Z2M also uses. 
If I scan the QR-Code on the side of the product with a regular QR-Code scanner I get the following (with added linebreaks for readability), my question will soon follow:
```
Z:F0D1B8000017389A
$I:6AB6973274EE6F720200530162754044C930
%M:1189
$D:008B452720
```
Out of the 4 parts, I already figured out the first 3. The "Z:" stands for Zigbee and the following 16 chars is just the MAC-Address. "I:" is the install code itself, followed by a 2 Byte CRC checksum. The "M%" is the manufacturer code, which I got from [here](https://update.ledvance.com/firmware-overview). 
Now I have no clue what the last "D:" stands for. Maybe the D is a device identifier of some sort? Googling the number with the name of the lightbulb yielded no results but maybe anyone here has an idea or can me help search?  
I have already contacted the LEDVANCE support and will let you know if they answer.  
I once again need to stress that I know that this is only loosely connected to Zigbee2MQTT but I believe it would be great if the functionality to add devices with a QR-Code could be added sometime in the future as this provides a better security because attackers cant easily sniff the network key during commissioning. 
Thanks for help and have a great day!

Edit. Forgot to mention that the Zigbee Specification doesn't define the QR-Code format which is why every manufacturer could basically implement it differently. 

Edit2. I will collect the QR-Code formats here:
- **LEDVANCE**:
```Z:<Zigbee EUI-64 (8 Byte)>$I:<Install Code (16 Byte) + CRC (2 Byte)>%M:<Manufacturer Code (2 Byte)$D:<Device Type Identifier (4 Byte)>```

- **Müller-Licht**:
```<Zigbee EUI-64 (8 Byte)>|<Install Code (16 Byte) + CRC (2 Byte)>```

- **Innr:**
```<Zigbee EUI-64 (8 Byte)>|<Install Code (16 Byte) + CRC (2 Byte)>```

**Comment by sjorge** [comment]
```
Z:F0D1B8000010B893$
I:A0C98017A0F75C55EF9512151815372640F6%
M:1189$
D:00110A5218
```

This is from the QR on an OSRAM Smart+ Classic E27 Multicolor, same manufacturer.

Thought some extra data points might help. I'm guessing the last one is either a device model identifier or maybe the default network address the device uses.

I noticed one my test Muller-Light Tint GU10's also has one:

```
000D6FFFFE63CCA0|58A4DD6123F38ECF22B17FDA0D95A43DED19
```

I'm guessing the format here is `MAC|INSTALLCODE`

**Reply by clemwo** [reply]
Thanks for your help. It made me realize that I also have a Müller Licht Set laying around that I havent even unboxed and to my surprise they also have QR-Codes. The format is the same as the one you posted. I also scanned another LEDVANCE product, the LEDVANCE Plug EU Smart+ Indoor and it has the same format as the light bulb.

**Reply by sjorge** [reply]
I'm guessing to get this to work probably only the mac and key are needed and the rest can be ignored.
At least they seem to be consistent within the same brand.

I also checked my desk light which is a Innr E14, it also has a QR:
```14B457FFFE2BD760|3B0D7FBE594092C5500FADF83AAA26E40F1D```

So this seems to follow the same format as Tint.

**Reply by clemwo** [reply]
You're probably right. The Bosch Smart Home App probably only uses the Manufacturer Code and the last "D" (I assume device type) to check whether the user has selected the correct device for joining and throws an error if its the wrong type. I just ordered the same LEDVANCE Bulb again in order to check whether the "D" value is the same.  
What I forgot to mention is that the "D" value was different in the smart plug compared to the light bulb, so it could be the device type.

**Comment by Koenkk** [comment]
FYI; the next z2m release will support install codes. It is already available in the dev branch (https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html) (frontend -> settings -> tools -> add install code).

Parsing logic: https://github.com/Koenkk/zigbee-herdsman/blob/7333692c774620b7eac7726a6f25110533ae1c59/src/controller/controller.ts#L214

**Reply by MartijnVdS** [reply]
The install code parsing logic does not handle the codes for LEDVANCE bulbs as described above in this discussion -- when I enter one, it expects 91 or 95 bytes instead.

**Comment by JohnConnett** [comment]
More information on **QR Code** install code formats [here](https://github.com/Koenkk/zigbee2mqtt/discussions/19982). This includes **Aqara**, **Bosch**, **Consciot**, **Danfoss**, **Enbrighten** and **Philips Hue**.
